File: j2me/c/roadmap_main.c
/* roadmap_main.c - The main function of the RoadMap application.
 *
 * LICENSE:
 *
 *   Copyright 2007 Ehud Shabtai
 *
 *   This file is part of RoadMap.
 *
 *   RoadMap is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or
 *   (at your option) any later version.
 *
 *   RoadMap is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with RoadMap; if not, write to the Free Software
 *   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * SYNOPSYS:
 *
 *   int main (int argc, char **argv);
 */

#include <java/lang.h>
#include <javax/microedition/lcdui.h>
#include <javax/microedition/lcdui/game.h>
#include <command_mgr.h>
#include <form_command_mgr.h>
#include <gps_manager.h>
#include <timer_mgr.h>
#include <device_specific.h>

#include <stdlib.h>
#include <string.h>

#include "roadmap.h"
#include "roadmap_path.h"
#include "roadmap_start.h"
#include "roadmap_config.h"
#include "roadmap_history.h"
#include "roadmap_canvas.h"
#include "roadmap_messagebox.h"
#include "roadmap_dialog.h"

#include "roadmap_main.h"

static RoadMapConfigDescriptor RoadMapConfigFullScreen =
                        ROADMAP_CONFIG_ITEM("Display", "Full Screen");

extern void roadmap_canvas_configure (void);

struct roadmap_main_io {
   int active;
   RoadMapIO io;
   RoadMapInput callback;
};

#define ROADMAP_MAX_IO 16
static struct roadmap_main_io RoadMapMainIo[ROADMAP_MAX_IO];


struct roadmap_main_timer {
   RoadMapCallback callback;
};

#define ROADMAP_MAX_TIMER 10
static struct roadmap_main_timer RoadMapMainPeriodicTimer[ROADMAP_MAX_TIMER];


static RoadMapKeyInput RoadMapMainInput = NULL;

volatile static int command_addr = 0;
volatile static int form_command_addr = 0;
volatile static char *form_command_name[255];
volatile static int form_command_context = 0;
volatile static int should_exit = 0;
static NOPH_GpsManager_t gps_mgr = 0;
static NOPH_TimerMgr_t timer_mgr = 0;
static NOPH_GameCanvas_t Canvas;
static const char *exception_type = "";
static int FullScreenMode = 1;

static void roadmap_start_event (int event) {
   switch (event) {
   case ROADMAP_START_INIT:
      break;
   }
}

static int KeyCode = 0;

static void roadmap_main_process_key (int keyCode) {

   char *k = NULL;
   const RoadMapAction *action;

   switch (keyCode) {
   case NOPH_Canvas_KEY_NUM2:
      k = "+";
      break;
   case NOPH_Canvas_KEY_NUM4:
      k = "J";
      break;
   case NOPH_Canvas_KEY_NUM5:
      k = "R";
      break;
   case NOPH_Canvas_KEY_NUM6:
      k = "K";
      break;
   case NOPH_Canvas_KEY_NUM8:
      k = "-";
      break;
   case NOPH_Canvas_KEY_POUND:
      KeyCode = 0;
      action = roadmap_start_find_action ("togglegpsrecord");
      if (action) (*action->callback) ();
      return;
   case NOPH_Canvas_KEY_STAR:
      k = "A";
      KeyCode = 0;
      break;
   default:
      switch (NOPH_Canvas_getGameAction((NOPH_Canvas_t)Canvas, keyCode)) {
      case NOPH_Canvas_UP:
         k = "Button-Up";
         break;
      case NOPH_Canvas_FIRE:
         KeyCode = 0;
         k = "G";
         break;
      case NOPH_Canvas_RIGHT:
         k = "Button-Right";
         break;
      case NOPH_Canvas_DOWN:
         k = "Button-Down";
         break;
      case NOPH_Canvas_LEFT:
         k = "Button-Left";
         break;
      }
   }

   //roadmap_log (ROADMAP_DEBUG, "In roadmap_main_process_key, keys:%d, k:%s, RoadMapMainInput:0x%x\n", keys, k, RoadMapMainInput);

   if ((k != NULL) && (RoadMapMainInput != NULL)) {
      (*RoadMapMainInput) (k);
   }
}


void roadmap_main_toggle_full_screen (void) {

   static int RoadMapIsFullScreen = 0;

   if (RoadMapIsFullScreen) {
   } else {
   }
}

void roadmap_main_new (const char *title, int width, int height) {
}


void roadmap_main_set_keyboard (struct RoadMapFactoryKeyMap *bindings,
                                RoadMapKeyInput callback) {
   RoadMapMainInput = callback;
}


RoadMapMenu roadmap_main_new_menu (void) {

   return NULL;
}


void roadmap_main_free_menu (RoadMapMenu menu) {

}


void roadmap_main_add_menu (RoadMapMenu menu, const char *label) {

}


void roadmap_main_add_menu_item (RoadMapMenu menu,
                                 const char *label,
                                 const char *tip,
                                 RoadMapCallback callback) {

   NOPH_CommandMgr_t cm = NOPH_CommandMgr_getInstance();
   NOPH_CommandMgr_addCommand(cm, label, (void *)callback);
}


void roadmap_main_popup_menu (RoadMapMenu menu, int x, int y) {

}


void roadmap_main_add_separator (RoadMapMenu menu) {

   roadmap_main_add_menu_item (menu, NULL, NULL, NULL);
}


void roadmap_main_add_tool (const char *label,
                            const char *icon,
                            const char *tip,
                            RoadMapCallback callback) {
}


void roadmap_main_add_tool_space (void) {

}


void roadmap_main_add_canvas (void) {
   RoadMapImage image;

   roadmap_config_declare_enumeration
      ("preferences", &RoadMapConfigFullScreen, NULL, "yes", "no", NULL);

   Canvas = NOPH_GameCanvas_get();

   if (!strcmp(roadmap_config_get (&RoadMapConfigFullScreen), "no"))
        FullScreenMode = 0;

   NOPH_Canvas_setFullScreenMode(Canvas, FullScreenMode);

   roadmap_log(ROADMAP_DEBUG, "In roadmap_main_add_canvas...\n");
   roadmap_canvas_configure ();

   roadmap_log(ROADMAP_DEBUG, "B4 load welcome...\n");
   image = roadmap_canvas_load_image (NULL, "/welcome.png");
   roadmap_log(ROADMAP_DEBUG, "After load welcome:%x\n", image);

   if (image) {
      RoadMapGuiPoint pos;

      RoadMapPen splash_bg = roadmap_canvas_create_pen ("SplashBG");
      roadmap_canvas_set_foreground ("#000000");
      roadmap_canvas_select_pen (splash_bg);

      roadmap_canvas_erase ();

      pos.x = (roadmap_canvas_width () - roadmap_canvas_image_width(image)) / 2;
      pos.y = (roadmap_canvas_height () - roadmap_canvas_image_height(image)) / 2;
      roadmap_canvas_draw_image (image, &pos, 0, IMAGE_NORMAL);
      roadmap_canvas_free_image (image);
      roadmap_canvas_refresh ();
   }
}


void roadmap_main_add_status (void) {

}


void roadmap_main_show (void) {

}

void roadmap_main_set_input (RoadMapIO *io, RoadMapInput callback) {

   if (io->subsystem == ROADMAP_IO_SERIAL) {
      /* We currently only support GPS input */
      if (!gps_mgr) {
         gps_mgr = NOPH_GpsManager_getInstance();
         NOPH_GpsManager_setTypeMsgs(gps_mgr,
                                     roadmap_lang_get("Internal GPS"),
                                     roadmap_lang_get("External GPS"));
      }
      RoadMapMainIo[0].io = *io;
      RoadMapMainIo[0].callback = callback;
      RoadMapMainIo[0].active = 1;
      NOPH_GpsManager_start(gps_mgr);
   }
}


void roadmap_main_remove_input (RoadMapIO *io) {

   if (io->subsystem == ROADMAP_IO_SERIAL) {
      /* We currently only support GPS input */
      if (gps_mgr) NOPH_GpsManager_stop(gps_mgr);
      RoadMapMainIo[0].active = 0;
   }
}


void roadmap_main_set_periodic (int interval, RoadMapCallback callback) {

   int index;

   if (!timer_mgr) timer_mgr = NOPH_TimerMgr_getInstance();
   index = NOPH_TimerMgr_set (timer_mgr, interval);

   if ((index == -1) || (RoadMapMainPeriodicTimer[index].callback != NULL)) {
      roadmap_log (ROADMAP_FATAL, "Can't create a new timer!");
   }


   RoadMapMainPeriodicTimer[index].callback = callback;
}


void roadmap_main_remove_periodic (RoadMapCallback callback) {

   int index;

   for (index = 0; index < ROADMAP_MAX_TIMER; ++index) {

      if (RoadMapMainPeriodicTimer[index].callback == callback) {

         RoadMapMainPeriodicTimer[index].callback = NULL;
         NOPH_TimerMgr_remove (timer_mgr, index);

         return;
      }
   }

   roadmap_log (ROADMAP_ERROR, "timer 0x%08x not found", callback);
}


void roadmap_main_set_status (const char *text) {

}


void roadmap_main_flush (void) {

}


void roadmap_main_exit (void) {

   int index;
   roadmap_log(ROADMAP_DEBUG, "b4 roadmap_start_Exit...\n");
   roadmap_start_exit ();
   roadmap_log(ROADMAP_DEBUG, "after roadmap_start_Exit...\n");
   should_exit = 1;

   /* remove all timers */
   for (index = 0; index < ROADMAP_MAX_TIMER; ++index) {

      if (RoadMapMainPeriodicTimer[index].callback != NULL) {

         RoadMapMainPeriodicTimer[index].callback = NULL;
         NOPH_TimerMgr_remove (timer_mgr, index);
      }
   }

   exit(0);
}

void roadmap_main_set_cursor (int cursor) {
   if (cursor == ROADMAP_CURSOR_WAIT) {

      if (roadmap_dialog_activate("Please wait", NULL, 1)) {
         roadmap_dialog_new_progress ("Please wait", "Please wait...");
         roadmap_dialog_complete (0);
      }

   } else if (cursor == ROADMAP_CURSOR_NORMAL) {
      roadmap_dialog_hide ("Please wait");
   }
}

static void keyPressed(int code)
{
   KeyCode = code;
   return;
}

static void keyReleased(int code)
{
  KeyCode = 0;
}

static void handler_main_loop(NOPH_Exception_t exception, void *arg)
{
  NOPH_Throwable_printStackTrace(exception);
  roadmap_messagebox(roadmap_lang_get("Error"), exception_type);

  NOPH_delete(exception);
}

//#define  DEBUG_TIME
static void wait_for_events(void)
{
#ifdef DEBUG_TIME
  int start_time = NOPH_System_currentTimeMillis();
  int end_time;
#endif

  while (!should_exit) {
     if (command_addr) break; /* Menu command */
     if (NOPH_FormCommandMgr_getCallBackNotif((int *)&form_command_addr, (void *)form_command_name, (void *)&form_command_context)) break;
     if (gps_mgr) {
#ifdef DEBUG_TIME
        int has_data = 0;
#endif
	exception_type = "GPS processing";
        while (RoadMapMainIo[0].active &&
              (NOPH_GpsManager_read(gps_mgr, 0, 0) != 0)) {
#ifdef DEBUG_TIME
           if (!has_data) {
              //roadmap_log(ROADMAP_DEBUG, "MAIN LOOP: got gps data...\n");
              has_data = 1;
           }
#endif
           /* GPS data is ready */
           RoadMapInput callback = RoadMapMainIo[0].callback;
           if (callback) (*callback)(&RoadMapMainIo[0].io);
        }
#ifdef DEBUG_TIME
        if (has_data) {
           end_time = NOPH_System_currentTimeMillis();
	   if ((end_time - start_time) > 50)
		roadmap_log(ROADMAP_DEBUG, "MAIN LOOP: finished processing GPS data in %d ms\n",
                  end_time - start_time);
           start_time = end_time;
        }
#endif
     }

     if (timer_mgr) {
        int index;
#ifdef DEBUG_TIME
        int has_timer = 0;
#endif
	exception_type = "Periodic timer";
        while ((index = NOPH_TimerMgr_getExpired(timer_mgr)) != -1) {
#ifdef DEBUG_TIME
           if (!has_timer) {
              //roadmap_log(ROADMAP_DEBUG, "MAIN LOOP: executing timer callbacks...\n");
              has_timer = 1;
           }
#endif
           (*RoadMapMainPeriodicTimer[index].callback) ();
        }
#ifdef DEBUG_TIME
        if (has_timer) {
           index = NOPH_System_currentTimeMillis();
	   if ((index - start_time) >= 50)
              roadmap_log(ROADMAP_DEBUG, "MAIN LOOP: finished callbacks in %d ms\n", index - start_time);
           start_time = index;
        }
#endif
     }

     if (KeyCode != 0) break;

     NOPH_Thread_sleep( 10 );
#ifdef DEBUG_TIME
     end_time = NOPH_System_currentTimeMillis();
     if ((end_time - start_time) > 100) roadmap_log(ROADMAP_DEBUG, "MAIN LOOP: Slept for %d ms!\n", end_time - start_time);
     start_time = end_time;
#endif
  }
}

#define SLEEP_PERIOD 100
int main (int argc, char **argv) {

   int i;

   NOPH_DeviceSpecific_init();

   NOPH_Canvas_registerKeyPressedCallback(keyPressed);
   NOPH_Canvas_registerKeyReleasedCallback(keyReleased);

   for (i = 0; i < ROADMAP_MAX_IO; ++i) {
      RoadMapMainIo[i].io.os.file = -1;
      RoadMapMainIo[i].io.subsystem = ROADMAP_IO_INVALID;
   }

   roadmap_start_subscribe (roadmap_start_event);
   roadmap_start (argc, argv);

   /*
   if (NOPH_exception) {
      NOPH_Throwable_printStackTrace(NOPH_exception);
      exit(1);
   }
*/
   NOPH_CommandMgr_setResultMem(NOPH_CommandMgr_getInstance(), (int *)&command_addr);
//   NOPH_FormCommandMgr_setCallBackNotif((int *)&form_command_addr, (void *)form_command_name, (void *)&form_command_context);

NOPH_try(handler_main_loop, NULL) {

   /* The main game loop */
   while(!should_exit)
   {
#ifdef DEBUG_TIME
      int start_time;
      int end_time;
#endif

      /* Wait for an event */
      wait_for_events();

#ifdef DEBUG_TIME
      start_time = NOPH_System_currentTimeMillis();
#endif
      if (command_addr) {
#ifdef DEBUG_TIME
         roadmap_log(ROADMAP_DEBUG, "MAIN LOOP: processing command...\n");
#endif
	 exception_type = "command";
         ((RoadMapCallback)command_addr)();
         command_addr = 0;
#ifdef DEBUG_TIME
         end_time = NOPH_System_currentTimeMillis();
         roadmap_log(ROADMAP_DEBUG, "MAIN LOOP: finished processing command in %d ms\n",
                  end_time - start_time);
         start_time = end_time;
#endif
      }

      if (form_command_addr) {
#ifdef DEBUG_TIME
         roadmap_log(ROADMAP_DEBUG, "MAIN LOOP: processing form command...\n");
         roadmap_log (ROADMAP_ERROR,
            "Form command: addr:%x, name:%s, context:%x",
            form_command_addr, form_command_name, form_command_context);
#endif
	 exception_type = "Form callback";
         ((RoadMapDialogCallback)form_command_addr)((char *)form_command_name,
                                          (void *)form_command_context);
         form_command_addr = 0;
#ifdef DEBUG_TIME
         end_time = NOPH_System_currentTimeMillis();
         roadmap_log(ROADMAP_DEBUG, "MAIN LOOP: finished processing form command in %d ms\n",
                  end_time - start_time);
         start_time = end_time;
#endif
      }

#ifdef DEBUG_TIME
      roadmap_log(ROADMAP_DEBUG, "MAIN LOOP: processing key command...\n");
#endif
      if (KeyCode) {
	 exception_type = "Key command";
         roadmap_main_process_key (KeyCode);
         NOPH_Thread_sleep( 30 );
      }
#ifdef DEBUG_TIME
      end_time = NOPH_System_currentTimeMillis();
      roadmap_log(ROADMAP_DEBUG, "MAIN LOOP: finished processing key command in %d ms\n",
            end_time - start_time);
      start_time = end_time;
#endif
   }
} NOPH_catch();

   roadmap_log(ROADMAP_DEBUG, "Main loop exiting...\n");
   //exit(0);
   return 0;
}

File: src/iphone/roadmap_main.m
/* roadmap_main.c - The main function of the RoadMap application.
 *
 * LICENSE:
 *
 *   Copyright 2002 Pascal F. Martin
 *   Copyright 2008 Morten Bek Ditlevsen
 *
 *   This file is part of RoadMap.
 *
 *   RoadMap is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or
 *   (at your option) any later version.
 *
 *   RoadMap is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with RoadMap; if not, write to the Free Software
 *   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * SYNOPSYS:
 *
 *   int main (int argc, char **argv);
 */

#include <stdlib.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>

#include "roadmap.h"
#include "roadmap_path.h"
#include "roadmap_start.h"
#include "roadmap_config.h"
#include "roadmap_history.h"
#include "roadmap_iphonecanvas.h"
#include "roadmap_iphonemain.h"

#include "roadmap_main.h"
#include "roadmap_time.h"

#import <Foundation/NSDictionary.h>

extern NSString *kUIButtonBarButtonAction;
extern NSString *kUIButtonBarButtonInfo;
extern NSString *kUIButtonBarButtonInfoOffset;
extern NSString *kUIButtonBarButtonSelectedInfo;
extern NSString *kUIButtonBarButtonStyle;
extern NSString *kUIButtonBarButtonTag;
extern NSString *kUIButtonBarButtonTarget;
extern NSString *kUIButtonBarButtonTitle;
extern NSString *kUIButtonBarButtonTitleVerticalHeight;
extern NSString *kUIButtonBarButtonTitleWidth;
extern NSString *kUIButtonBarButtonType;


static RoadMapApp *TheApp;
static int RoadMapMainToolbarAdded = 0;
static int sArgc;
static char ** sArgv;

struct roadmap_main_io {
   NSFileHandle *fh;
   RoadMapIO io;
   RoadMapInput callback;
};

#define ROADMAP_MAX_IO 16
static struct roadmap_main_io RoadMapMainIo[ROADMAP_MAX_IO];

struct roadmap_main_timer {
   NSTimer *timer;
   RoadMapCallback callback;
};

#define ROADMAP_MAX_TIMER 16
static struct roadmap_main_timer RoadMapMainPeriodicTimer[ROADMAP_MAX_TIMER];

static RoadMapCallback idle_callback = NULL;
static char *RoadMapMainTitle = NULL;

static RoadMapKeyInput RoadMapMainInput = NULL;
static UIWindow    *RoadMapMainWindow  = NULL;
static UIView      *RoadMapMainBox     = NULL;
static RoadMapCanvasView *RoadMapCanvasBox   = NULL;
//static UIView      *RoadMapMainMenuBar = NULL;
static UIButtonBar *RoadMapMainToolbar = NULL;
//static UIView      *RoadMapMainStatus  = NULL;
static NSArray     *RoadMapMainToolbarArray = NULL;
static int          buttonCount = 0;
static RoadMapCallback RoadMapMainToolbarCallbacks[6];


static char *roadmap_main_toolbar_icon (const char *icon) {
    unsigned int i;

    if (icon == NULL)
        return NULL;

    const char *icon_file = roadmap_path_search_icon (icon);
    if (icon_file == NULL)
        return NULL;

    for (i = 0; i < strlen(icon_file); i++)
    {
        if (strncmp (icon_file + i, "resources", 9) == 0)
            return (char *)(icon_file + i);
    }
    return NULL;
}

void roadmap_main_toggle_full_screen (void) {
   /* Don't care abobut full screen on the iPhone */
}

void roadmap_main_new (const char *title, int width, int height) {
    [TheApp newWithTitle: title andWidth: width andHeight: height];
}

void roadmap_main_title(char *fmt, ...) {
   /* Don't care about titles on the iPhone */
}

void roadmap_main_set_keyboard (RoadMapKeyInput callback) {
   RoadMapMainInput = callback;
}

RoadMapMenu roadmap_main_new_menu (const char *title) {
   return NULL;
}


void roadmap_main_free_menu (RoadMapMenu menu) {
     NSLog (@"roadmap_main_free_menu\n");
}


void roadmap_main_add_menu (RoadMapMenu menu, const char *label) {
     NSLog (@"roadmap_main_add_menu label: %s\n", label);
/*
   UIView *menu_item;

   if (RoadMapMainMenuBar == NULL) {

      RoadMapMainMenuBar = gtk_menu_bar_new();

      gtk_box_pack_start
         (GTK_BOX(RoadMapMainBox), RoadMapMainMenuBar, FALSE, TRUE, 0);
   }

   menu_item = gtk_menu_item_new_with_label (label);
   gtk_menu_shell_append (GTK_MENU_SHELL(RoadMapMainMenuBar), menu_item);

   gtk_menu_item_set_submenu (GTK_MENU_ITEM(menu_item), (UIView *) menu);
*/
}


void roadmap_main_popup_menu (RoadMapMenu menu,
                              const RoadMapGuiPoint *position) {
     NSLog (@"roadmap_main_popup_menu\n");

 /*  if (menu != NULL) {
      gtk_menu_popup (GTK_MENU(menu),
                      NULL,
                      NULL,
                      NULL,
                      NULL,
                      0,
                      gtk_get_current_event_time());
   }
   */
}


void roadmap_main_add_menu_item (RoadMapMenu menu,
                                 const char *label,
                                 const char *tip,
                                 RoadMapCallback callback) {

     NSLog (@"roadmap_main_add_menu_item label: %s tip: %s\n", label, tip);
/*
   UIView *menu_item;

   if (label != NULL) {

      menu_item = gtk_menu_item_new_with_label (label);
      g_signal_connect (menu_item, "activate",
                        (GCallback)roadmap_main_activate,
                        callback);
   } else {
      menu_item = gtk_menu_item_new ();
   }
   gtk_menu_shell_append (GTK_MENU_SHELL(menu), menu_item);
   gtk_widget_show(menu_item);

   if (tip != NULL) {
      gtk_tooltips_set_tip (gtk_tooltips_new (), menu_item, tip, NULL);
   }
   */
}


void roadmap_main_add_separator (RoadMapMenu menu) {
   NSLog (@"roadmap_main_add_seperator\n");

   roadmap_main_add_menu_item (menu, NULL, NULL, NULL);
}


void roadmap_main_add_toolbar (const char *orientation) {
     NSLog (@"roadmap_main_add_toolbar orientation: %s\n", orientation);
     RoadMapMainToolbarAdded = 1;
     RoadMapMainToolbarArray = [[NSArray array] init];
}

void roadmap_main_add_tool (const char *label,
                            const char *icon,
                            const char *tip,
                            RoadMapCallback callback) {
    char *iconstr;
    NSString *nsicon, *nslabel;

    NSLog (@"roadmap_main_add_tool label: %s icon: %s tip: %s\n", label, icon, tip);
    if (buttonCount >= 5)
    {
        NSLog (@"roadmap_main_add_tool only room for 5 buttons\n");
        return;
    }
    buttonCount++;

    NSArray *tmp = RoadMapMainToolbarArray;
    nslabel = [[NSString alloc] initWithUTF8String:icon];
    iconstr = roadmap_main_toolbar_icon(icon);
    if (iconstr)
        nsicon = [[NSString alloc] initWithUTF8String:iconstr];
    else
        nsicon = @"";

    RoadMapMainToolbarCallbacks[buttonCount] = callback;
    NSDictionary *dict = [ NSDictionary dictionaryWithObjectsAndKeys:
        @"buttonBarItemTapped:", kUIButtonBarButtonAction,
        nsicon, kUIButtonBarButtonInfo,
        @"imagedown.png", kUIButtonBarButtonSelectedInfo,
        [ NSNumber numberWithInt: buttonCount], kUIButtonBarButtonTag,
        TheApp, kUIButtonBarButtonTarget,
        nslabel, kUIButtonBarButtonTitle,
        //@"", kUIButtonBarButtonTitle,
        @"0", kUIButtonBarButtonType,
        nil
    ];
    RoadMapMainToolbarArray = [tmp arrayByAddingObject: dict];
}


void roadmap_main_add_tool_space (void) {
    /* Don't care about spaces on the iPhone... */
}

static unsigned long roadmap_main_busy_start;

void roadmap_main_set_cursor (RoadMapCursor newcursor) {
    /* Don't care about cursors on the iPhone... */
}

void roadmap_main_busy_check(void) {

   if (roadmap_main_busy_start == 0)
      return;

   if (roadmap_time_get_millis() - roadmap_main_busy_start > 1000) {
      roadmap_main_set_cursor (ROADMAP_CURSOR_WAIT);
   }
}

void roadmap_main_add_canvas (void) {
    struct CGRect rect = [UIHardware fullScreenApplicationContentRect];
    rect.origin.x = 0.0f;
    rect.origin.y = 0.0f;
    if (RoadMapMainToolbarAdded)
       rect.size.height = 480.0f - 49.0f - 20.0f;
    else
       rect.size.height = 480.0f - 20.0f;

    RoadMapCanvasBox = [[RoadMapCanvasView alloc] initWithFrame: rect];
    [RoadMapMainBox addSubview: RoadMapCanvasBox];
}

void roadmap_main_add_status (void) {
   /* Status bar seems like a waste of space on the iPhone... */
}


void roadmap_main_show (void) {
    /* Since this is called after the toolbar is
       configured, we use this to add the finalized
       toolbar to the main view */
     if (RoadMapMainToolbarAdded)
     {
        RoadMapMainToolbar = [TheApp createButtonBar];
        [RoadMapMainBox addSubview: RoadMapMainToolbar]; 
     }
}

void roadmap_main_set_input (RoadMapIO *io, RoadMapInput callback) {
    [TheApp setInput: io andCallback: callback];
}

void roadmap_main_remove_input (RoadMapIO *io) {
    [TheApp removeInput: io];
}

void roadmap_main_set_periodic (int interval, RoadMapCallback callback) {
    [TheApp setPeriodic: (interval*0.001) andCallback: callback];
}


void roadmap_main_remove_periodic (RoadMapCallback callback) {
   [TheApp removePeriodic: callback];
}


void roadmap_main_set_status (const char *text) {
//   NSLog (@"roadmap_main_set_status text: %s\n", text);
}

void roadmap_main_set_idle_function (RoadMapCallback callback) {
    idle_callback = callback;
    [TheApp setPeriodic: 0.00 andCallback: callback];
}

void roadmap_main_remove_idle_function (void) {
   [TheApp removePeriodic: idle_callback];
   idle_callback = NULL;
}


int roadmap_main_flush (void) {
   if ([[NSNotificationCenter defaultCenter] isEmpty])

      return 0;

   return 1;
//   NSLog (@"roadmap_main_flush\n");
/*
   while (gtk_events_pending ()) {
      if (gtk_main_iteration ()) {
         exit(0);  // gtk_main_quit() called 
      }
   }
   */
   return 0;
}


int roadmap_main_flush_synchronous (int deadline) {
   NSLog (@"roadmap_main_flush_synchronous\n");

   long start_time, duration;

   start_time = roadmap_time_get_millis();

  /* while (gtk_events_pending ()) {
      if (gtk_main_iteration ()) {
         exit(0); 
      }
   }
   */
//   gdk_flush();

   duration = roadmap_time_get_millis() - start_time;

   if (duration > deadline) {

      roadmap_log (ROADMAP_DEBUG, "processing flush took %d", duration);

      return 0; /* Busy. */
   }

   return 1; /* Not so busy. */
}

void roadmap_main_exit (void) {

   static int exit_done;

   if (!exit_done++) {
      roadmap_start_exit ();
   }
}

int main (int argc, char **argv) {
    int i;
    int j = 0;
    sArgc = argc;
    sArgv = (char **)malloc(argc * (sizeof (char*)));
    for (i=0; i<argc; i++)
    {
        if (strcmp(argv[i], "--launchedFromSB") != 0) {
            sArgv[i] = strdup(argv[j]);
            j++;
        }
        else
           sArgc--;
    }
    putenv("HOME=/var/mobile/Library/");

    // Add paths to roadmap drivers and to flite
    putenv("PATH=/Applications/RoadMap.app/bin:/usr/bin");
    
    roadmap_option (sArgc, sArgv, 0, NULL);
    NSAutoreleasePool *pool = [[NSAutoreleasePool alloc] init];
    return UIApplicationMain(sArgc, sArgv, [RoadMapApp class]);
    [pool release];
    return 0;
}

@implementation RoadMapApp

- (UIButtonBar *)createButtonBar {
    UIButtonBar *buttonBar;
    buttonBar = [ [ UIButtonBar alloc ]
       initInView: RoadMapMainBox
       withFrame: CGRectMake(0.0f, 480.0f - 49.0f - 20.0f, 320.0f, 49.0f)
       withItemList: RoadMapMainToolbarArray ];
    [buttonBar setDelegate:self];
    [buttonBar setBarStyle:0];
    [buttonBar setButtonBarTrackingMode: 1];

    int buttons[5] = { 1, 2, 3, 4, 5};
    [buttonBar registerButtonGroup:0 withButtons:buttons withCount: buttonCount];
    [buttonBar showButtonGroup: 0 withDuration: 0.0f];
    int tag;

    for(tag = 1; tag <= buttonCount; tag++) {
        [ [ buttonBar viewWithTag:tag ]
            setFrame:CGRectMake(2.0f + ((tag - 1) * 64.0f), 1.0f, 64.0f, 48.0f)
        ];
    }

    return buttonBar;
}

- (void)buttonBarItemTapped:(id) sender {
    int button = [ sender tag ];
    printf("buttonclicked: %i\n", button);
      (RoadMapMainToolbarCallbacks[button] ) ();
}

-(void) newWithTitle: (const char *)title andWidth: (int) width andHeight: (int) height
{
    if (RoadMapMainWindow == NULL) {
        struct CGRect rect = [UIHardware fullScreenApplicationContentRect];
        rect.origin.x = rect.origin.y = 0.0f;
        RoadMapMainWindow = [[UIWindow alloc] initWithContentRect: rect];
        RoadMapMainBox = [[UIView alloc] initWithFrame: rect];
        [RoadMapMainWindow orderFront: self];
        [RoadMapMainWindow makeKey: self];
        [RoadMapMainWindow _setHidden: NO];
        [RoadMapMainWindow setContentView: RoadMapMainBox];
    }

    if (RoadMapMainTitle != NULL) {
        free(RoadMapMainTitle);
    }
    RoadMapMainTitle = strdup (title);
}

- (void) periodicCallback: (NSTimer *) timer
{
   int i;
   for (i = 0; i < ROADMAP_MAX_TIMER; ++i) {
      if (RoadMapMainPeriodicTimer[i].timer == timer) {
         (* RoadMapMainPeriodicTimer[i].callback) ();
         break;
      }
   }
}

-(void) setPeriodic: (float) interval andCallback: (RoadMapCallback) callback
{
   int index;
   struct roadmap_main_timer *timer = NULL;

   for (index = 0; index < ROADMAP_MAX_TIMER; ++index) {

      if (RoadMapMainPeriodicTimer[index].callback == callback) {
         return;
      }
      if (timer == NULL) {
         if (RoadMapMainPeriodicTimer[index].callback == NULL) {
            timer = RoadMapMainPeriodicTimer + index;
         }
      }
   }

   if (timer == NULL) {
      roadmap_log (ROADMAP_FATAL, "Timer table saturated");
   }

   timer->callback = callback;
   timer->timer = [NSTimer scheduledTimerWithTimeInterval: interval
                     target: self
                     selector: @selector(periodicCallback:)
                     userInfo: nil
                     repeats: YES];
}

-(void) removePeriodic: (RoadMapCallback) callback
{
   int index;

   for (index = 0; index < ROADMAP_MAX_TIMER; ++index) {

      if (RoadMapMainPeriodicTimer[index].callback == callback) {

         RoadMapMainPeriodicTimer[index].callback = NULL;
         [RoadMapMainPeriodicTimer[index].timer invalidate];
         return;
      }
   }
}

-(void) ioCallback: (id) notify
{
   NSFileHandle *fh = [notify object];
   int i;
   int fd = [fh fileDescriptor];

   for (i = 0; i < ROADMAP_MAX_IO; ++i) {
      if (RoadMapMainIo[i].io.os.file == fd) {
         (* RoadMapMainIo[i].callback) (&RoadMapMainIo[i].io);
            [RoadMapMainIo[i].fh waitForDataInBackgroundAndNotify];
         break;
      }
   }

}

-(void) setInput: (RoadMapIO*) io andCallback: (RoadMapInput) callback
{
    int i;
    int fd = io->os.file; /* All the same on UNIX. */
    NSFileHandle *fh = [[NSFileHandle alloc] initWithFileDescriptor: fd];

    for (i = 0; i < ROADMAP_MAX_IO; ++i) {
        if (RoadMapMainIo[i].io.subsystem == ROADMAP_IO_INVALID) {
            RoadMapMainIo[i].io = *io;
            RoadMapMainIo[i].callback = callback;
            RoadMapMainIo[i].fh = fh;
            [[NSNotificationCenter defaultCenter]
                addObserver: self
                selector:@selector(ioCallback:)
                name:NSFileHandleDataAvailableNotification
                object:fh];
            [fh waitForDataInBackgroundAndNotify];
            break;
        }
    }
}

-(void) removeInput: (RoadMapIO*) io
{
    int i;
    int fd = io->os.file; /* All the same on UNIX. */

    for (i = 0; i < ROADMAP_MAX_IO; ++i) {
        if (RoadMapMainIo[i].io.os.file == fd) {
            [[NSNotificationCenter defaultCenter]
                removeObserver: self
                name:NSFileHandleDataAvailableNotification
                object:RoadMapMainIo[i].fh];
            RoadMapMainIo[i].io.os.file = -1;
            RoadMapMainIo[i].io.subsystem = ROADMAP_IO_INVALID;
            break;
        }
    }
}

- (void) applicationDidFinishLaunching: (id) unused
{
    TheApp = self;
    int i;
    for (i = 0; i < ROADMAP_MAX_IO; ++i) {
        RoadMapMainIo[i].io.os.file = -1;
        RoadMapMainIo[i].io.subsystem = ROADMAP_IO_INVALID;
    }

    roadmap_start (sArgc, sArgv);

    [self reportAppLaunchFinished];
}

- (void)applicationWillSuspend
{
 //  [self terminate];
   printf("go to sleep\n");
}

- (void)applicationDidResume
{
  printf("I am alive!\n");

}

- (void)applicationWillTerminate
{
    roadmap_main_exit();
    [TheApp release];
}

@end

